### PR TITLE
feat(pipeline): support pipeline plan

### PR DESCRIPTION
**Pull request by warren830**
## 

blueprint

Summary

modify create pipeline to support pipeline plan
create patch to enable/disable pipeline plan

Users can set schedule on config-ui, then frontend will parse the user input to crontab format and send back to backend.
Please check cron time format in https://crontab.guru/
One more thing, if it's interval, the format should like this "*/5 * * * ", add '' before '/'

Important

As this pr changed id&createdAt&updatedAt api output to camelCase, this will impact "task" in frontend

Other Information

In the last commit, I divided api pipelines to pipeplines and pipelianplans as blow:
r.GET("/pipelines", pipelines.Index)
r.GET("/pipelines/:pipelineId", pipelines.Get)
r.GET("/blueprints", blueprints.Index)
r.GET("/blueprints/:blueprintId", blueprints.Get)
r.POST("/pipelines", pipelines.Post)
r.POST("/blueprints", blueprints.Post)
r.DELETE("/pipelines/:pipelineId", pipelines.Delete)
r.DELETE("/blueprints/:blueprintId", blueprints.Delete)
r.PATCH("/blueprints/:blueprintId", blueprints.PUT)
r.GET("/pipelines/:pipelineId/tasks", task.Index)
r.GET("/ping", ping.Get)
r.POST("/push/:tableName", push.Post)

POST /blueprints

    Request
    {
    	"name": "COLLECT BLUEPRINT",
    	"tasks": [
    		[
    			{
    				"plugin": "github",
    				"options": {
    					"repo": "lake",
    					"owner": "merico-dev"
    				}
    			}
    		]
    	],
    	"enable": true,
    	"cronConfig": "103 13 /13 * *"
    }
    Response
    {
    	"id": 7,
    	"name": "COLLECT BLUEPRINT",
    	"tasks": [
    		[
    			{
    				"plugin": "github",
    				"options": {
    					"owner": "merico-dev",
    					"repo": "lake"
    				}
    			}
    		]
    	],
    	"enable": true,
    	"cronConfig": "103 13 /13 * *",
    	"createdAt": "2022-03-27T10:16:20.046+08:00",
    	"updatedAt": "2022-03-27T10:16:20.046+08:00"
    }

GET /blueprints

    Response
    {
    	"count": 2,
    	"blueprints": [
    		{
    			"name": "COLLECT BLUEPRINT",
    			"tasks": [
    				[
    					{
    						"plugin": "github",
    						"options": {
    							"repo": "lake",
    							"owner": "merico-dev"
    						}
    					}
    				]
    			],
    			"enable": true,
    			"cronConfig": "*/1 * * * *",
    			"id": 9,
    			"createdAt": "2022-03-28T12:31:00.117Z",
    			"updatedAt": "2022-03-28T12:31:00.117Z"
    		},
    		{
    			"name": "COLLECT BLUEPRINT",
    			"tasks": [
    				[
    					{
    						"plugin": "github",
    						"options": {
    							"repo": "lake",
    							"owner": "merico-dev"
    						}
    					}
    				]
    			],
    			"enable": false,
    			"cronConfig": "",
    			"id": 7,
    			"createdAt": "2022-03-27T02:16:20.046Z",
    			"updatedAt": "2022-03-27T02:16:20.046Z"
    		}
    	]
    }

GET /blueprints/:blueprintId

    Response
    {
    	"id": 7,
    	"name": "COLLECT BLUEPRINT",
    	"tasks": [
    		[
    			{
    				"plugin": "github",
    				"options": {
    					"owner": "merico-dev",
    					"repo": "lake"
    				}
    			}
    		]
    	],
    	"enable": true,
    	"cronConfig": "103 13 /13 * *",
    	"createdAt": "2022-03-27T10:16:20.046+08:00",
    	"updatedAt": "2022-03-27T10:16:20.046+08:00"
    }

PUT /blueprints/:blueprintId

    Request
    {
    	"name": "COLLECT BLUEPRINT",
    	"tasks": [
    		[
    			{
    				"plugin": "github",
    				"options": {
    					"repo": "lake",
    					"owner": "merico-dev"
    				}
    			}
    		]
    	],
    	"enable": true, // this param must be passed 
    	"cronConfig": "103 13 /13 * *"
    }
    Response
    {
    	"id": 7,
    	"name": "COLLECT BLUEPRINT",
    	"tasks": [
    		[
    			{
    				"plugin": "github",
    				"options": {
    					"owner": "merico-dev",
    					"repo": "lake"
    				}
    			}
    		]
    	],
    	"enable": true,
    	"cronConfig": "103 13 /13 * *",
    	"createdAt": "2022-03-27T10:16:20.046+08:00",
    	"updatedAt": "2022-03-27T10:16:20.046+08:00"
    }

### CODE REVIEW

File: services/pipeline.go [hunk 1]
```diff
@@ -256,3 +291,36 @@ func CancelPipeline(pipelineId uint64) error {
 	}
 	return err
 }
+
+func ModifyPipelinePlan(newPipeline *models.NewPipeline) (*models.PipelinePlan, error) {
+	cronConfig := newPipeline.CronConfig
+	cronTime := ""
+	switch newPipeline.CronConfig.Type {
```
Comment (klesh) [inline services/pipeline.go:298]: we should save/load whatever user input, and only do this kind of parsing right before usage.

File: services/pipeline.go [hunk 2]
```diff
@@ -31,7 +32,38 @@ func init() {
 	db.Model(&models.Pipeline{}).Where("status = ?", models.TASK_RUNNING).Update("status", models.TASK_FAILED)
 }
 
-func CreatePipeline(newPipeline *models.NewPipeline) (*models.Pipeline, error) {
+func CreatePipelinePlan(newPipeline *models.NewPipeline) (*models.PipelinePlan, error) {
+	cronConfig := newPipeline.CronConfig
+	cronTime := ""
+	switch newPipeline.CronConfig.Type {
```
Comment (klesh) [inline services/pipeline.go:38]: same as below

**Comment by klesh** [comment]
Please use camelCase for all out api output, we are going to remove all PascalCase in near future.

File: models/pipeline_run.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+// to be executed concurrently, while each set is to be executed sequentially.
+type NewPipeline struct {
+	Name           string       `json:"name"`
+	PipelinePlanId uint64       `json:"pipeline_plan_id"`
+	Tasks          [][]*NewTask `json:"tasks"`
+	CronConfig     *CronConfig
+	Enable         string
+}
+
+type CronConfig struct {
```
[the remark this answers is not in the document]
Comment (warren830) [reply models/pipeline_run.go:34]: pipelin_plan.go is another file which save pipeline_plan.go, this file saves pipeline and newPipeline as before

[the remark this answers is not in the document]
Comment (warren830) [reply models/pipeline_run.go:34]: I reverted the file to pipeline.go

File: models/pipeline_plan.go [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+	"github.com/merico-dev/lake/models/common"
+	"gorm.io/datatypes"
+)
+
+type PipelinePlan struct {
+	common.Model
+	Name     string
+	CronTime string         //cron job format, like '5 * * * 2'
+	Tasks    datatypes.JSON `json:"tasks"`
+	Enable   string
```
[the remark this answers is not in the document]
Comment (warren830) [reply models/pipeline_plan.go]: fixed

File: models/pipeline_run.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (warren830) [reply models/pipeline_run.go]: And I moved CronConfig to pipeline_plan.go in the new commit

File: services/pipeline.go (hunk 1 above)
Comment (warren830) [reply services/pipeline.go]: Fixed, sorry, I misunderstood in the first commit

**Comment by warren830** [comment]
In the last commit, I divided api pipelines to pipeplines and pipelianplans as blow:
	r.GET("/pipelines", pipelines.Index)
	r.GET("/pipelines/:pipelineId", pipelines.Get)
	r.GET("/pipelineplans", pipelineplans.Index)
	r.GET("/pipelineplans/:pipelinePlanId", pipelineplans.Get)
	r.POST("/pipelines", pipelines.Post)
	r.POST("/pipelineplans", pipelineplans.Post)
	r.DELETE("/pipelines/:pipelineId", pipelines.Delete)
	r.DELETE("/pipelineplans/:pipelinePlanId", pipelineplans.Delete)
	r.PATCH("/pipelineplans/:pipelinePlanId", pipelineplans.Patch)
	r.GET("/pipelines/:pipelineId/tasks", task.Index)
	r.GET("/ping", ping.Get)
	r.POST("/push/:tableName", push.Post)

**Comment by warren830** [comment]
GET /pipelineplans
```json
{
	"id": 16,
	"createdAt": "2022-03-27T10:06:55.535+08:00",
	"updatedAt": "2022-03-27T10:06:55.552+08:00",
	"name": "COLLECT 1648121282469",
	"pipelinePlanId": 6,
	"tasks": [
		[
			{
				"plugin": "github",
				"options": {
					"owner": "merico-dev",
					"repo": "lake"
				}
			}
		]
	],
	"totalTasks": 1,
	"finishedTasks": 0,
	"beganAt": null,
	"finishedAt": null,
	"status": "TASK_CREATED",
	"message": "",
	"spentSeconds": 0
}
```

**Comment by warren830** [comment]
GET /pipelineplans
```json
Response
{
	"id": 7,
	"createdAt": "2022-03-27T10:16:20.046+08:00",
	"updatedAt": "2022-03-27T10:16:20.046+08:00",
	"name": "COLLECT 1648121282469",
	"tasks": [
		[
			{
				"plugin": "github",
				"options": {
					"owner": "merico-dev",
					"repo": "lake"
				}
			}
		]
	],
	"enable": true,
	"type": "interval",
	"day": 13,
	"hour": 13,
	"minute": 103
}
```

**Comment by warren830** [comment]
GET /pipelineplans/:pipelinePlanId
```json
Response
{
	"id": 7,
	"createdAt": "2022-03-27T10:16:20.046+08:00",
	"updatedAt": "2022-03-27T10:16:20.046+08:00",
	"name": "COLLECT 1648121282469",
	"tasks": [
		[
			{
				"plugin": "github",
				"options": {
					"owner": "merico-dev",
					"repo": "lake"
				}
			}
		]
	],
	"enable": true,
	"type": "interval",
	"day": 13,
	"hour": 13,
	"minute": 103
}
```

**Comment by warren830** [comment]
PATCH /pipelineplans/:pipelinePlanId
```json
Request
{
	"name": "COLLECT 1648121282469",
	"tasks": [
		[
			{
				"plugin": "github",
				"options": {
					"repo": "lake",
					"owner": "merico-dev"
				}
			}
		]
	],
	"enable": true,
	"cronConfig": {
	    "type": "interval",
	    "day": 13,
	    "hour": 13,
	    "minute": 103,
	}
}
Response
{
	"id": 7,
	"createdAt": "2022-03-27T10:16:20.046+08:00",
	"updatedAt": "2022-03-27T10:16:20.046+08:00",
	"name": "COLLECT 1648121282469",
	"tasks": [
		[
			{
				"plugin": "github",
				"options": {
					"owner": "merico-dev",
					"repo": "lake"
				}
			}
		]
	],
	"enable": true,
	"type": "interval",
	"day": 13,
	"hour": 13,
	"minute": 103
}
```

**Comment by klesh** [comment]
TODOs:

- [ ] add cronConfig document link to explain how it works
- [ ] implementation(devlake would schedule existing plans on start up, and adjust dynamically when any plan changed via api request)
- [ ] add documentation for pipeline plan api
- [ ] remove lake-cli and its documentation
- [ ] config-ui support pipeline plan

File: services/blueprint.go [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+	return nil
+}
+
+func ReloadBlueprints() error {
+	blueprints := make([]*models.Blueprint, 0)
+	err := db.Model(&models.Blueprint{}).Where("enable = ?", true).Find(&blueprints).Error
+	if err != nil {
+		panic(err)
+	}
+	c := cron.New()
```
Comment (klesh) [inline services/blueprint.go:115]: We should clean up old cron jobs before loading new job. Local variable doesn't seem right

File: services/blueprint.go [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+		newPipeline.BlueprintId = pp.ID
+		_, err = c.AddFunc(pp.CronConfig, func() {
+			pipeline, err := CreatePipeline(&newPipeline)
+			// Return all created tasks to the User
+			if err != nil {
+				cLog.Error("created cron job failed: %s", err)
+				return
+			}
+			go func() {
+				_ = RunPipeline(pipeline.ID)
```
[the remark this answers is not in the document]
Comment (warren830) [reply services/blueprint.go:145]: I just copied and pasted old code...fixed

File: api/blueprints/blueprints.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+	"github.com/merico-dev/lake/models"
+	"github.com/merico-dev/lake/services"
+)
+
+func Post(c *gin.Context) {
+	inputBlueprint := &models.InputBlueprint{}
+
+	err := c.MustBindWith(inputBlueprint, binding.JSON)
+	if err != nil {
+		logger.Error("post /blueprint failed", err)
```
Comment (klesh) [inline api/blueprints/blueprints.go:21]: Please avoid using logger.Error, it was deprecated.

File: api/blueprints/blueprints.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+	blueprintId := c.Param("blueprintId")
+	id, err := strconv.ParseUint(blueprintId, 10, 64)
+	if err != nil {
+		shared.ApiOutputError(c, err, http.StatusBadRequest)
+		return
+	}
+	editBlueprint := &models.EditBlueprint{}
+	err = c.MustBindWith(editBlueprint, binding.JSON)
+	if err != nil {
+		logger.Error("patch /blueprints/:blueprintId failed", err)
```
Comment (klesh) [inline api/blueprints/blueprints.go:88]: We dont need to log it, because we can see error from response anyway

File: services/blueprint.go [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+	return &blueprint, nil
+}
+
+func DeleteBlueprint(id uint64) error {
+	err := db.Delete(&models.Blueprint{}, "id = ?", id).Error
+	if err != nil {
+		logger.Error("create pipline failed", err)
+		return errors.InternalError
+	}
+	return nil
```
Comment (klesh) [inline services/blueprint.go:111]: We need to reload cron jobs here

File: services/blueprint.go [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+	}
+	for _, pp := range blueprints {
+		var tasks [][]*models.NewTask
+		err = json.Unmarshal(pp.Tasks, &tasks)
+		if err != nil {
+			cLog.Error("created cron job failed: %s", err)
+			return err
+		}
+		//
+		newPipeline := models.NewPipeline{}
```
Comment (klesh) [inline services/blueprint.go:133]: newPipeline should be created inside the cron callback, or concurrency error might occur.

File: services/blueprint.go [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+				cLog.Error("run cron job failed: %s", err)
+				return
+			}
+			cLog.Info("Run new cron job successfully")
+		})
+		if err != nil {
+			cLog.Error("created cron job failed: %s", err)
+			return err
+		}
+		err = db.Create(&models.CronEntry{
```
Comment (klesh) [inline services/blueprint.go:155]: Will this entryId still valid after restart? it make no sense to save it into db if it didn't.

File: services/blueprint.go [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+	err = json.Unmarshal(blueprint.Tasks, &tasks)
+	if err != nil {
+		cLog.Error("created cron job failed: %s", err)
+		return err
+	}
+	newPipeline := models.NewPipeline{}
+	newPipeline.Tasks = tasks
+	newPipeline.Name = blueprint.Name
+	newPipeline.BlueprintId = blueprint.ID
+	entryId, err := c.AddFunc(blueprint.CronConfig, func() {
```
Comment (klesh) [inline services/blueprint.go:199]: I still prefer that we compose a `reloadBlueprints`, which will kill all existing cronjobs, and then create new cronjobs based on what we have in database, that is simpler, and enough.
and call this function after Create/Update/Delete operations.

Comment (warren830) [reply services/blueprint.go:199]: fixed

File: services/blueprint.go (hunk 11 above)
Comment (warren830) [reply services/blueprint.go:155]: fixed

File: services/blueprint.go (hunk 10 above)
Comment (warren830) [reply services/blueprint.go:133]: Yes, you are right!

File: api/blueprints/blueprints.go (hunk 8 above)
Comment (warren830) [reply api/blueprints/blueprints.go:88]: fixed

File: api/blueprints/blueprints.go (hunk 7 above)
Comment (warren830) [reply api/blueprints/blueprints.go:21]: fixed

File: api/blueprints/blueprints.go [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+		shared.ApiOutputError(c, err, http.StatusBadRequest)
+		return
+	}
+	err = services.DeleteBlueprint(id)
+	if err != nil {
+		shared.ApiOutputError(c, err, http.StatusBadRequest)
+	}
+}
+
+func PUT(c *gin.Context) {
```
[the remark this answers is not in the document]
Comment (warren830) [reply api/blueprints/blueprints.go:78]: fixed

**Review by klesh — APPROVED** [review]
LGTM

**Pull request merged**